### PR TITLE
Fix a contact rollups PD calculation and refactor large query

### PR DESCRIPTION
**Pull request by jeremydstone**
This PR fixes/improves a professional development enrollment calculation. One table links enrollments to user ID (which has an associated email) and/or a standalone email not necessarily associated with an account. Both may be set and they are not always set to the same thing. With this fix, PD enrollment will be marked for either for both emails.

This PR also breaks up a large query which had gotten unwieldy. Fundamentally, we need to look for presence of 6 courses in 4 different tables and build a single comma-separated string per email address summarizing the results. Previously, `update_professional_learning_attendance` had done this in one large query, which involved UNIONS of 4 different subqueries for the 4 different sources. The advantage of doing it all at once is we could just build the result in one place and not need to worry about appending to existing result values the database. The disadvantage is the query got really hard to understand and it wasn't possible to run one part in isolation while developing. I broke it into a multi-pass approach where for each of the 6 courses we individually analyze the 4 sources and append any results to the output if we need to. It is easier to read and runs overall almost as quickly.

### CODE REVIEW

**Comment by ashercodeorg** [comment]
Approximately how many rows are involved in this giant refactor? As it, it still remains largely incomprehensible. Is it possible to grab the relevant rows in a `SELECT` query, manipulate it in Ruby, and update the appropriate rows in an `UPDATE` query?

**Comment by jeremydstone** [comment]
The row count involved in this particular aspect of the rollup is not huge - around 50K. For this aspect, it would be possible to rewrite as a Ruby row-by-row method. That's not practical for pretty much every other part of the process, where the row count is 6 or 7 figures. This is ETL and the SQL you see here is a pretty standard pattern for ETL.

**Comment by ashercodeorg** [comment]
Yep, I realize SQL is a standard pattern for ETL. But this particular SQL is barely comprehensible, I'm looking for possible methods to make the rollup more readable, more maintainable, and more testable.

File: lib/cdo/contact_rollups.rb [hunk 1]
```diff
@@ -39,9 +39,19 @@ class ContactRollups
...
   # need to filter here to known courses in the Pardot schema - we can't blindly pass values through
-  COURSE_LIST = "'CS Fundamentals','CS in Algebra','CS in Science','CS Principles','Exploring Computer Science','CS Discoveries'".freeze
+  COURSE_ARRAY = [
+    "CS Fundamentals",
+    "CS in Algebra",
+    "CS in Science",
+    "CS Principles",
+    "Exploring Computer Science",
+    "CS Discoveries"
+  ].freeze
```
Comment (ashercodeorg) [inline lib/cdo/contact_rollups.rb:51]: Can these course names be referenced from somewhere that defines these strings as constants? E.g., from dashboard/app/models/pd/workshop.rb?

**Comment by ashercodeorg** [comment]
LGTM, though I think we are descending further and further into the realm of unreadability, untestability, unmaintainability.

**Pull request merged**